src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -1074,20 +1262,69 @@ struct BlockZipperJoinImpl {
     compatibleRowAction_.flush();
   }
 
-  // The actual join routine that combines all the previous functions.
+  // Helper function for `addRemainingUndefPairs`. Process blocks that have not
+  // been consumed yet and match them against undef blocks.
+  template <bool reversed>
+  bool consumeRemainingBlocks(auto& side, const auto& undefBlocks) {
+    bool hasUndef = false;
+    while (side.it_ != side.end_) {
+      const auto& lBlock = *side.it_;
+      for (const auto& rBlock : undefBlocks) {
+        if constexpr (reversed) {
+          compatibleRowAction_.setInput(rBlock.fullBlock(), lBlock);
+        } else {
+          compatibleRowAction_.setInput(lBlock, rBlock.fullBlock());
+        }
+        for (size_t i : ad_utility::integerRange(lBlock.size())) {
+          for (size_t j : rBlock.getIndexRange()) {
+            hasUndef = true;
+            if constexpr (reversed) {
+              compatibleRowAction_.addRow(j, i);
+            } else {
+              compatibleRowAction_.addRow(i, j);
+            }
+          }
+        }
+      }
+      ++side.it_;
+    }
```

Comment:
Okay, I am stopping the review for now.
I am not yet fully sure why the undef support takes three times the code of the old version.
Maybe we can use a design similar to the fully materialized zipper join for bla with undef where the undef handling is completely templated out.